src/main/scala/org/broadinstitute/dsde/workbench/sam/util/LdapSupport.scala

```diff
@@ -24,31 +29,36 @@ trait LdapSupport {
     * @tparam T
     * @return
     */
-  protected def ldapSearchStream[T](baseDn: String, searchScope: SearchScope, filters: Filter*)(unmarshaller: Entry => T): Stream[T] =
-    filters.flatMap { filter =>
-      val search = new SearchRequest(baseDn, searchScope, filter)
-      val entrySource = new LDAPEntrySource(ldapConnectionPool.getConnection, search, true)
-      ldapEntrySourceStream(entrySource)(unmarshaller)
-    }.toStream
-
-  // this is the magic recursive stream generator
-  private def ldapEntrySourceStream[T](entrySource: LDAPEntrySource)(unmarshaller: Entry => T): Stream[T] =
-    Try(Option(entrySource.nextEntry)) match {
-      case Success(None) =>
-        // reached the last element, Stream.empty terminates the stream
-        Stream.empty
-
-      case Success(Some(next)) =>
-        // next element exists, return a Stream starting with unmarshalled next followed by the rest of the stream
-        // (streams are smart and lazily evaluate the second parameter)
-        Stream.cons(unmarshaller(next), ldapEntrySourceStream(entrySource)(unmarshaller))
-
-      case Failure(ldape: EntrySourceException)
-          if ldape.getCause.isInstanceOf[LDAPException] &&
-            ldape.getCause.asInstanceOf[LDAPException].getResultCode == ResultCode.NO_SUCH_OBJECT =>
-        Stream.empty // the base dn does not exist, treat as empty search
-      case Failure(regrets) => throw regrets
+  protected def ldapSearchStream[T](baseDn: String, searchScope: SearchScope, filters: Filter*)(unmarshaller: Entry => IO[T]): fs2.Stream[IO, T] = {
+    for {
+      filter <- fs2.Stream(filters:_*)
+      entry <- ldapSearch(baseDn, searchScope, filter)
+      unmarshalled <- fs2.Stream.eval(unmarshaller(entry))
+    } yield {
+      unmarshalled
+    }
+  }
+
+  private def ldapSearch(baseDn: String, searchScope: SearchScope, filter: Filter): fs2.Stream[IO, Entry] = {
+    val search = new SearchRequest(baseDn, searchScope, filter)
+    for {
+      entrySource <- fs2.Stream.eval(executeLdap(IO(
+        new LDAPEntrySource(ldapConnectionPool.getConnection, search, true))))
+
+      entry <- fs2.Stream.unfoldEval[IO, LDAPEntrySource, Entry](entrySource) { entrySourceInner =>
+        nextEntry(entrySourceInner).map(_.map(entry => (entry, entrySourceInner)))
+      }
+    } yield entry
+  }
+
+  private def nextEntry(entrySource: LDAPEntrySource): IO[Option[Entry]] = {
+    executeLdap(IO(Option(entrySource.nextEntry))).handleErrorWith {
```

Comment:
this is the gist of why we want fs2 here: every call to entrySource.nextEntry can block and that needs to be shifted to a different context